### ISSUE TITLE
Why the spectral_radius can be 0?

### ISSUE THREAD

**Issue by imhithanks** [TUD-STKS/PyRCN#8]
1) Note Onset Detection using Echo State Networks http://www.essv.de/pdf/2020_157_164.pdf
In this paper , spectral radius is optimized in the range from 0.0 to 1.5.
Why the spectral_radius can be 0?
The reservoir matrix Wres was normalized by its largest absolute eigenvalue to achieve a spectral
radius ρ = 1.0.
I think Wres is multipled by the spectral_radius we choose so that Wres's ρ=the the spectral_radius we choose.
If we set  spectral_radius=0, will Wres=0? 

2) reservoir_size in old version =hidden_layer_size in new version,is it right?

3) k_res determines the number of nodes that every node inside the reservoir is connected with.
I think k_res=(1-sparsity rate)*hidden_layer_size, is it right?

4) In many examples , k_res has not been optimized. Is k_res not important?
Shoud It be optimized with hidden_layer_size together using a sequence of grid and line searches?
For example, I set k_res=0.1*hidden_layer_size, k_res=0.2*hidden_layer_size, ... k_res=1*hidden_layer_size
and find the best k_res. Is it right?

Thank you for your help.

**Comment by renierts** [comment]
Well, if the spectral_radius is zero, it simply means that we don't have any recurrent connections. After normalizing Wres to its largest absolute eigenvalue, you can multiply Wres with a factor, the same way as input_scaling and Win.

2. Yes, that's true.
3. k_res is another way to describe the sparsity. Assume that hidden_layer_size = 100 and k_rec =10: Every neuron in the reservoir will receive information from 10 randomly selected other neurons in the reservoir. Thus, sparsity=k_rec / hidden_layer_size
4. You can do that. However, if you read papers of Jalalvand et al. (e.g. http://sibgrapi.sid.inpe.br/col/sid.inpe.br/sibgrapi/2015/07.31.20.05/doc/Aza_SIBGRAPI2015.pdf), you will find out that k_rec has only minor impact on the performance of an ESN.

**Comment by imhithanks** [comment]
@renierts I see. Thank you so much！！

**Comment by imhithanks** [comment]
Dear @renierts，
I 'd like to ask you some more questions.
1) wash_out: This element defines the number of first samples to be not regarded for training.
In many examples , wash_out=0. Is wash_out important?
if I have 1000 samples, how much should wash_out be setted generally?Is it necessary to optimize wash_out？
2) I  find that some people use optimal algorithms to optimize the parameters，such as PSO and Genetic Algorithm.
Is your strategy of optimization（Optimize input scaling and spectral radius→Optimize leakage and bias→Optimize beta→Increase reservoir size)beter than the optimal algorithms ?
I think your strategy is easier. I'd like to hear your opinion.
Thank you for your help.

**Comment by renierts** [comment]
Hi @imhithanks ,
I have rarely used wash_out and implemented it more for the sake of completeness. For time series prediction, it might be useful to optimize it, but you should try that out.

Yes, there are many different ways. I think, our strategy is easier, as many other optimization algorithms might need a lot of to determine the best set of parameters. Furthermore, the dimensionality of the search space (4 hyper-parameters or more) might be too large for, e.g. Bayes Optimization.

**Comment by imhithanks** [comment]
@renierts Oh I see.
I am using the old version PyRCN.  I found that k_in and k_res cannot be 1.
k_in=-1 means k_res=1.
k_res=-1 means  k_res=1.
is it right?

**Comment by renierts** [comment]
Thanks for notifying this. I think, this was a bug. Actually, it should have been possible to set k_in=1 and k_res=1.

In your case with a one-dimensional input, you are right. k_in=-1 and k_res=-1 means dense connections. In the current version, I have removed the -1.

**Comment by imhithanks** [comment]
I see .Thank you very much.

**Comment by imhithanks** [comment]
Dear @renierts，
I'd like to ask you some questions again.
Multiple-F0 Estimation using Echo State Networks:
 https://www.eurasip.org/Proceedings/Eusipco/Eusipco2020/pdfs/0000126.pdf
Note Onset Detection using Echo State Networks: http://www.essv.de/pdf/2020_157_164.pdf

According to the papers above, 
All values in Win and Wres were initialized from a standard normal distribution.
Wbi consists of fixed random values from a uniform distribution between ±1.0.

1)  Why you use different distributions to initialize the 3 matrixes?
2) Can Win  and Wres be initialized from a uniform distribution between ±1.0?

**Comment by renierts** [comment]
Hi @imhithanks ,

In PyRCN, you cannot change the distribution from outside right now. However, both normal and uniform distributions to initialized the 3 matrices have been used.

I used a uniform distribution to have clear upper and lower limits for the bias scaling. Right now, I have even switched to uniformly distributed input weights to also have definitely defined upper and lower limits for the input scaling.

As Wres is further normalized to a unitary spectral radius, I see no benefit in initializing Wres from a uniform distribution.

**Comment by imhithanks** [comment]
Thank you so much. And I want to know more about it.
1)In PyRCN0.0.4,Win were initialized from a standard normal distribution. Is it true?
In PyRCN0.0.6,Win were initialized from a uniformly distribution. Is it true?
2)Do different initialization methods of Win have a big impact on the result in a time series forecasting task?

**Comment by renierts** [comment]
1. No, Win was always initialized from a standard normal distribution in PyRCN. The first papers you refer to were published before the initial release of PyRCN. During the development, we decided to use a uniform distribution for the input weights.
2. Different Random initializations do not have a big impact on my classification results. I would guess, the impact on time series forecasting should be minor.

**Comment by imhithanks** [comment]
I see. And I have some more questions.
I always set random_state = 1.
1)When I repeat the experiment，the results often change with the same parameters .
I have used the same random_state.Why do the results still change?
2) How can I make sure that the result remains the same when I repeat the experiment.

**Comment by renierts** [comment]
This is strange, because it works for me all the time.

Could you please give me an example?

**Comment by renierts** [comment]
Dear @imhithanks, is this still an open issue or can it be closed?

**Comment by renierts** [comment]
I close this issue for now.
